NEW show alert when subsite changes in another tab

NightJar opened the pull request:
Adds a new feature that alerts a user who is editing a site in the same browser but over multiple tabs, and changes the subsite in one tab. Normally nothing untoward would happen until the user then tries to save data for a subsite that is no longer the current one in session. Now a javascript powered modal will pop up alerting users to the fact that the subsite has changed, and includes a button to revert in order to continue editing (so changes aren't lost).

Resolves #314 
Closes #411

Tested working in IE11.

![image](https://user-images.githubusercontent.com/778003/66449966-50f83480-eab3-11e9-8904-e0472fca1bf0.png)

ichaber committed:
Setup as base silverstripe-module

ichaber committed:
Add basics to store subsiteID in localStorage.

ichaber committed:
Add package-lock.json

ichaber committed:
Get the build running and remove unnecessary files.

Commit:
NEW react component to block editing when subsite changes

When a content editor is working in 2 tabs, and they change the active subsite in one, the second tab will now block further editing with a modal notifying them they cannot continue editing until they change the subsite back in the tab they changed it in :)

ichaber committed:
Merge pull request #1 from creative-commoners/fix/411-multitab-subsite

NEW react component to block editing when subsite changes

Commit:
NEW revert button that matches active subsite to tab

The subsite ID is updated in the PHP session, and the JavaScript had no way of
reverting this back once if a user saw the modal on in their admin session. This
could lead to confusion about what to do next. Now there is a revert button that
will set the PHP session SubsiteID back to that of the tab beneath the modal,
making it safe to edit once again.

In order to support this, other tabs also need to update, and modals either displayed
or hidden, which was already a feature. However these changes only happen on remote tabs
(or Documents as per the API definition), so a new custom event has been introduced in
order to be able to have the current tab/document update itself - but without coupling
the React component to jQuery.entwine that handles the component lifecycle (externally
from React).

Commit:
Added context to messages show to user

Such as the subsite that the user is working on in that tab and the subsite that they are changing to

Commit:
FIX Don't double up subsite change notices

If using multiple subsites and more than two tabs you can create multiple react notices by changing the subsite more than once to different subsites in the same tab
To fix this we check for an open notice and always remove it first

ichaber committed:
Merge pull request #2 from creative-commoners/fix/411-multitab-subsite

NEW revert button that matches active subsite to tab

ichaber committed:
Merge pull request #3 from Mossman1215/fix/411-multitab-subsite

Fix/411 multitab subsite

Commit:
Trigger SubsiteID update on load, safely

Previous attempts to update the stored Subsite ID in the localStorage when a page loaded for the first time would cause an infinite refresh loop.
Reasons to set the ID on load are for hard refreshed tabs were a way to remove the modal blocking an edit, and could cause sync issues when multiple tabs are open - undermining the efforts recent commits have set to resolve.

This also solves an issue where loading a edit form for a page that did not exist on any other subsite would set the SubsiteID in the PHP Session backend, based on the subsite the page in question. This would create multiple tab sync issues in the CMS silently (no way to tell the UI to block an edit). By setting the SubsiteID when the page loads fresh we can avoid these situations.

Commit:
Refactor after testing and adjusting for edge cases

Not much functional change, mostly tidy up after testing that updates to the localStorage about the SubsiteID were made only after a successful switch from one subsite to the next (as opposed to assuming success, and possibly getting into the de-sync state that this work seeks to prevent).

Commit:
Fix subsite change notice working in IE 11

IE 11 has limited support for Event and StorageEvents, and needs some help
to create and fire the inforation at the correct timings. This commit sees
changes made to ensure the correct information is always received by the
handlers, triggering the notice appropriately.

Commit:
Resolve JS and CSS linting issues

Preventing the produciton build of front end assets from completing.
This is largely white space changes, but also includes a few minor
code adjustments around the JS to satisfy the linter - such as
Object.keys(obj).foreach((val) => ...) instead of for(val in obj).

Commit:
Add tests for new SubsiteChangeAlert javascript

This lead to refactoring the SubsiteChangeAlert component to not rely
on global state of the DOM, and instead take props in and reference them
directly to perform its required tasks. Equally it accepts a callback for
delegating the revert task. This is cleaner and more testable.

Also fix small issue with adding divs for the modal as many times as it
appears.

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+/* jslint browser: true, nomen: true */
```
seems like this file should have a .jsx extension

NightJar replied:
This is not the Silverstripe way. All supported modules use .js for react components as well as other types of JS.
I actually agree, but consistency is a nice thing.

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+
+  $('.cms-edit-form').entwine({
+    /**
+     * TODO: Fix with Entwine API extension. See https://github.com/silverstripe/silverstripe-subsites/pull/125
```
is this TODO going to get done?

NightJar replied:
No, this was existing code, moved in order to facilitate the new yarn build chain to support React components.
https://github.com/silverstripe/silverstripe-subsites/blob/fd10b868ec5dfccd8bd0273c92a30892e2af48cb/javascript/LeftAndMain_Subsites.js#L66 (current `2` branch HEAD)

emteknetnz replied:
So, what you're saying is this TODO is just copy pasted from existing code

The link points to a PR merged in 2013, which is not really helpful

If so, just remove the TODO? It's currently just confusing

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+        const href = $(this).attr('href');
+
+        if (typeof href !== 'undefined' && !href.match(/^http:\/\//)) {
+          $(this).attr('href', $.path.addSearchParams(href, {
```
will this add a duplicate search/querystring param if SubsiteID already exists?

NightJar replied:
Existing code outside scope of work for this PR
https://github.com/silverstripe/silverstripe-subsites/blob/fd10b868ec5dfccd8bd0273c92a30892e2af48cb/javascript/LeftAndMain_Subsites.js#L120

But no, it updates the param if it exists.

![image](https://user-images.githubusercontent.com/778003/75292010-3965f080-5888-11ea-8feb-80d15f4a0b58.png)

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+      $(doc).find('form').each(() => {
+        const action = $(this).attr('action');
+
+        if (typeof action !== 'undefined' && !action.match(/^http:\/\//)) {
```
also needs to match on https

NightJar replied:
This is existing code, but seems like a bug we can easily fix. OK.

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+        const action = $(this).attr('action');
+
+        if (typeof action !== 'undefined' && !action.match(/^http:\/\//)) {
+          $(this).append(`<input type=hidden name="SubsiteID" value="${subsiteId}" >`);
```
type="hidden"

NightJar replied:
Existing code, but we did refactor it to a backtick string without noticing that, good catch!

emteknetnz commented on client/src/legacy/entwine/SubsitesTreeDropdownField.js:
```diff
@@ -0,0 +1,33 @@
+   */
+  $('select.subsitestreedropdownfield-chooser').entwine({
+    onchange() {
+      // TODO Data binding between two fields
```
will this TODO get done?

NightJar replied:
Not in this PR.
Existing code, current head of `2` branch:
https://github.com/silverstripe/silverstripe-subsites/blob/fd10b868ec5dfccd8bd0273c92a30892e2af48cb/javascript/SubsitesTreeDropdownField.js#L9

emteknetnz commented on client/src/legacy/entwine/SubsitesTreeDropdownField.js:
```diff
@@ -0,0 +1,33 @@
+
+      const attributes = this._super();
+      attributes.data.urlTree += `?${fieldName}_SubsiteID=${subsiteID}`;
+      attributes.data.cacheKey = `${attributes.data.cacheKey.substring(0, 19)}_${subsiteID}`;
```
why are we getting a .substring(0, 19) ?

NightJar replied:
Because Entwine 🤷‍♂ 
Existing code, not altered in this PR - just moved into a new folder structure.
https://github.com/silverstripe/silverstripe-subsites/blob/fd10b868ec5dfccd8bd0273c92a30892e2af48cb/javascript/SubsitesTreeDropdownField.js#L27

emteknetnz commented on client/src/lib/changeActiveSubsite.js:
```diff
@@ -0,0 +1,20 @@
+import createEvent from 'legacy/createEvent';
+
+export default function changeActiveSubsite(toSubsiteID, toSubsiteName) {
+  const { localStorage } = window;
```
can get rid of this line and just use window.localStorage below

emteknetnz commented on client/src/components/SubsiteChangeAlert/tests/SubsiteChangeAlert-test.js:
```diff
@@ -0,0 +1,44 @@
+  });
+  describe('getMessage', () => {
+    it('should show the old subsite name correctly', () => {
+      const alert = shallow(
```
rename this since window.alert() is an existing function, so pretty confusing have a variable with the same name

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+      // React business
+      const modalContainer = this.getModalNode() || document.createElement('div');
+      document.body.appendChild(modalContainer);
+      const ChangeAlert = loadComponent('SubsiteChangeAlert');
```
should be `const SubsiteChangeAlert` so that the jsx below has the correct component name?

emteknetnz commented on client/src/components/SubsiteChangeAlert/SubsiteChangeAlert.jsx:
```diff
@@ -0,0 +1,59 @@
+    this.handleRevert = this.handleRevert.bind(this);
+  }
+
+  handleRevert() {
```
I'm finding the use of the word 'revert' throughout this confusing.  Shouldn't this be 'change' instead? (or click?)

NightJar replied:
No, revert describes what the intent is, not what the event is.
Can you think of a better suggestion that describes a users intent? We are struggling - longer forms still using "revert" in a more descriptive manner are all that's coming up.

emteknetnz commented on client/src/legacy/createEvent.js:
```diff
@@ -0,0 +1,20 @@
+/* global window */
+// Polyfill IE11
+function createEvent(type, extraData) {
+  const { document, Event } = window;
```
what does this line do?  document and Event are already globally available

NightJar replied:
Makes it very clear where these come from.
It's a small file so could they can be shifted into the global declaration atop the file though I suppose.

emteknetnz commented on client/src/legacy/createEvent.js:
```diff
@@ -0,0 +1,20 @@
+    event = new Event(type);
+  }
+  if (extraData) {
+    Object.keys(extraData).forEach((key) => {
```
`forEach(key =>`

NightJar replied:
🤔 Interesting, I'm sure we did this because the linter was complaining, however it doesn't seem to be the case. It does bring a question of code style and consistency into play though...
E.g. I think the inclusion of parentheses makes `someCall('firstParam', firstParam => {...})` a little more readable.

emteknetnz commented on client/src/legacy/createEvent.js:
```diff
@@ -0,0 +1,20 @@
+/* global window */
+// Polyfill IE11
```
this doesn't seem like it's a polyfill, might be better to remove this comment and put a comment in the `new Event(type)` block that this is for IE11

NightJar replied:
Changed comment to state that it is a helper function instead.

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+    ModalNode: null,
+    onadd() {
+      // Storage has updated - another tab has changed the subsite info
+      window.addEventListener('storage', (storageEvent) => {
```
`storage` is possibly a bit too generic, `subsitestorage` is more likely to never have a conflict

Mossman1215 replied:
Yeah that's the actual api name of the event type we're using.
https://developer.mozilla.org/en-US/docs/Web/API/StorageEvent
https://developer.mozilla.org/en-US/docs/Web/API/Storage

NightJar replied:
Indeed, this is not determined by us. It must be 'storage'.

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+import createEvent from 'legacy/createEvent';
+import changeActiveSubsite from 'lib/changeActiveSubsite';
+
+jQuery.entwine('ss', ($) => {
```
`$ =>` instead of `($) =>` - there's multiple instances where this can be updated, though I'll just mention it once here

NightJar replied:
As with https://github.com/silverstripe/silverstripe-subsites/pull/443#discussion_r384174797 I'll not make a change until this discussion reaches an outcome.

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+      window.localStorage.setItem('subsiteInfo', JSON.stringify(subsiteInfo));
+    },
+    showReactiveNotice(newSubsiteName) {
+      const { document } = window;
```
don't need this line, document is already available

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+      $(doc).find('a').each(() => {
+        const href = $(this).attr('href');
+
+        if (typeof href !== 'undefined' && !href.match(/^http:\/\//)) {
```
need to match on https as well

NightJar replied:
Seems like a duplicate comment for some reason 🤔 https://github.com/silverstripe/silverstripe-subsites/pull/443#pullrequestreview-349089514

NightJar replied:
Oh, not they're different lines 😨 
Done.

emteknetnz commented on client/src/legacy/entwine/LeftAndMain_Subsites.js:
```diff
@@ -0,0 +1,190 @@
+     * This is done so we can use the CMS domain for displaying previews so we prevent single-origin
+     * violations and SSL cert problems that come up when iframing content from a different URL.
+     */
+    onafterIframeAdjustedForPreview(event, doc) {
```
where is this called from? there's no other reference to this function in this PR

NightJar replied:
Entwine.
I imagine it's not even in this module.

emteknetnz commented on client/src/lib/changeActiveSubsite.js:
```diff
@@ -0,0 +1,20 @@
+    subsiteID: toSubsiteID,
+    subsiteName: toSubsiteName
+  };
+  request.open('GET', `?SubsiteID=${toSubsiteID}`);
```
what is this actually doing? will this "refresh the whole CMS"?
is this something that's already done here? https://github.com/silverstripe/silverstripe-subsites/pull/443/files#diff-cf163c29c1a03e751d8439e2f36b5206R33

NightJar replied:
No, it won't update the CMS in terms of what the user sees, but it does "update the CMS" in regards of what the server keeps in session.

The line you referenced (`window.location.search = $.query.set('SubsiteID', subsiteID);`) does reload the CMS in the browser - this by contrast sends an XHR that updates the server and only fires update events on completion, meaning CMS users do not lose their draft.

This file should have this info in a docblock. We'll add one.

emteknetnz commented on package.json:
```diff
@@ -0,0 +1,74 @@
+  "main": "./client/src/boot/index.js",
+  "author": "You",
+  "engines": {
+    "node": ">=10.x"
```
probably want to lock this to ^10 like other core packages e.g. https://github.com/silverstripe/silverstripe-asset-admin/blob/1/package.json#L17

NightJar replied:
Probably correct, although this was originally copied from another supported module, things change over time :)

emteknetnz commented on client/src/styles/LeftAndMain_Subsites.css:
```diff
@@ -0,0 +1,82 @@
+  top: 0;
+  width: 350px;
+  text-align: right;
+  margin-right: 130px;
```
margin: 0 130px 0 0;

NightJar replied:
No, as this would override any other margin applied to those other dimensions. This specifically sets _only_ the right margin.

Also not a part of this PR. File is simply moved from original location.
https://github.com/silverstripe/silverstripe-subsites/blob/2/css/LeftAndMain_Subsites.css#L12

emteknetnz commented on client/src/styles/LeftAndMain_Subsites.css:
```diff
@@ -0,0 +1,82 @@
+  margin-left: -100px;
+}
+
+body.SubsiteAdmin .right form #URL .fieldgroup * {
```
this selector looks a bit interesting, would it be possible to put a class="myclass" on something instead?

NightJar replied:
Also not a part of this PR. File is simply moved from original location.
https://github.com/silverstripe/silverstripe-subsites/blob/2/css/LeftAndMain_Subsites.css#L67

I assume this has to do with specificity interplaying with other CMS UI rules.

emteknetnz commented on package.json:
```diff
@@ -0,0 +1,74 @@
+    "css": "WEBPACK_CHILD=css npm run build",
+    "test": "jest",
+    "coverage": "jest --coverage",
+    "lock": "npm-shrinkwrap --dev",
```
shouldn't need shrinkwrap for anything since we are using yarn?

NightJar replied:
Probably correct. I'll remove it and see what happens.
Most amusingly it appears that shrinkwrap isn't listed as a dependency anyway 😂

emteknetnz commented on client/src/styles/LeftAndMain_Subsites.css:
```diff
@@ -0,0 +1,82 @@
+}
+
+.cms-add-form #PageType li .class-SubsitesVirtualPage {
+  background-position: 0 -32px !important;
```
would it be possible to do this without using !important?

NightJar replied:
Also not a part of this PR. File is simply moved from original location.
https://github.com/silverstripe/silverstripe-subsites/blob/2/css/LeftAndMain_Subsites.css#L72

brynwhyman commented:
Hey @Cheddam you made a mention that you've been working on this recently. Do you have some local changes you could push up?

Cheddam commented:
@brynwhyman They're not quite ready to push, but I'll try and make time to get them polished off tomorrow.

Commit:
Carry out code cleanliness tasks

Such as putting in docblocks to inform future developers what the
intent is behind a particular piece of behaviour. Linting has also
taken place, and pre-existing minor bugs fixed in legacy code.

Commit:
FIX Improve messaging, tidy up logic and syntax

Cheddam commented:
I've resolved (as in clicked 'resolve' on) all of the feedback items, as they were either addressed or refuted. I've also pushed some additional tidy-up and tweaks that I made while reviewing this (only took me 2 months to get them up, sorry team 😅).

Final steps:

- [ ] Squash to one commit, ensuring all contributors are listed
- [ ] Rebase on `2` branch (this isn't a breaking change, so not sure why it's pointed at `master` - any thoughts @NightJar?)
- [ ] Merge and release

emteknetnz commented on client/src/components/SubsiteChangeAlert/SubsiteChangeAlert.jsx:
```diff
@@ -0,0 +1,79 @@
+/* global window */
```
This file extension needs to be changed to `.js` to match the rest of the codebase

emteknetnz commented on client/src/components/SubsiteChangeAlert/SubsiteChangeAlert.jsx:
```diff
@@ -0,0 +1,79 @@
+
+    return i18n.inject(
+      i18n._t(
+        'SubsiteChangeAlert.SUBSITE_CHANGED',
```
should this be added to `en.yml`?

emteknetnz commented on client/src/components/SubsiteChangeAlert/SubsiteChangeAlert.jsx:
```diff
@@ -0,0 +1,79 @@
+        newSubsiteName,
+        currentSubsiteName
+      }
+    ).split('[SPLIT]');
```
It would be cleaner to return an array with two translatable strings `SUBSITE_CHANGED_LINE_ONE` and `SUBSITE_CHANGED_LINE_TWO`

emteknetnz commented on client/src/legacy/entwine/SubsitesTreeDropdownField.js:
```diff
@@ -0,0 +1,33 @@
+   */
+  $('select.subsitestreedropdownfield-chooser').entwine({
+    onchange() {
+      // TODO Data binding between two fields
```
Please do something with this TODO, either remove it (preferable) or add some context to the comment e.g. linking it to something relevant so that future developers know what it is or when it can be safely removed

emteknetnz reviewed (changes requested):
It looks good, though a few small changes should be made

emteknetnz committed:
Rename SubsiteChangeAlert.jsx to SubsiteChangeAlert.js

emteknetnz committed:
Update LeftAndMain_Subsites.js

emteknetnz commented:
Behat failure is the preview/split mode in using the main site domain instead of the subsite domain.

Re-ran travis `2` https://travis-ci.org/github/silverstripe/silverstripe-subsites/builds/640118109 which is green

Have tried rebasing on top of the latest `2` locally and still getting same behat failure

Different issue - got this when running `yarn install & yarn build` (have already done the same for admin module)

![image](https://user-images.githubusercontent.com/4809037/87539559-36ae4f80-c6f2-11ea-99ea-abc4b47d61cd.png)

emteknetnz commented:
Yeah, I'm not feeling this one as it is

I find the behat test failure to pretty concerning because it seems to show we've created a new layer in which to create werid bugs.

For instance, why do we need to do things like this? https://github.com/silverstripe/silverstripe-subsites/pull/443/files#diff-cf163c29c1a03e751d8439e2f36b5206R176

It seems as though we now need to hunt around and find all the things that now need to be updated via javascript just so that it works with this, new way of doing things?  I'm concerned that we're adding complexity and brittleness just to support a nice-to-have though not exactly essential feature.

Subsites already has a reputation for having weird bugs.  This PR will probably make things worse.

As it is I am not confident merging this. I'd like to see a decent chunk of this PR removed and simplified.  It may even need change of approach to get there.

Mossman1215 commented:
> Yeah, I'm not feeling this one as it is
> 
> I find the behat test failure to pretty concerning because it seems to show we've created a new layer in which to create werid bugs.
> 
> For instance, why do we need to do things like this? https://github.com/silverstripe/silverstripe-subsites/pull/443/files#diff-cf163c29c1a03e751d8439e2f36b5206R176
> 
> It seems as though we now need to hunt around and find all the things that now need to be updated via javascript just so that it works with this, new way of doing things? I'm concerned that we're adding complexity and brittleness just to support a nice-to-have though not exactly essential feature.
> 
> Subsites already has a reputation for having weird bugs. This PR will probably make things worse.
> 
> As it is I am not confident merging this. I'd like to see a decent chunk of this PR removed and simplified. It may even need change of approach to get there.

It was supposed to be the simplest way possible to fix the issue where the user is editing in two tabs between subsites and changes their php session subsite id from what is in their tab cache.
it leads to issues where users are putting pages in the wrong subsite and raising tickets.

I think it's ok to close this PR and I'll think about a better approach.

emteknetnz commented:
@Mossman1215 I think the idea and general approach here is a good one and it's worth fixing.  The issue is that this particuarly javascript implementation is overly complicated and that's creates a maintenance problem on a module that's already complicated.

I'll close this one for now, though likely that we'll either reopen this in the future, or, start a new PR and reuse/reference the work done in this one
